internal/attestation/untyped.go

```diff
@@ -0,0 +1,161 @@
+// Copyright The Enterprise Contract Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package attestation
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/in-toto/in-toto-golang/in_toto"
+	"github.com/sigstore/cosign/v2/pkg/oci"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/enterprise-contract/ec-cli/internal/signature"
+	"github.com/enterprise-contract/ec-cli/internal/utils"
+)
+
+// Some important keys we're hoping to find in the opaque data
+const (
+	predicateTypeKey = "predicateType"
+	predicateKey     = "predicate"
+)
+
+// Similar function to in_toto.ProvenanceStatementSLSA02 but we'll use
+// an opaque untyped map
+type UntypedAttestation map[string]any
+
+// UntypedProvenanceFromSignature parses an attestation layer from the provided OCI
+// layer. Expects that the layer contains DSSE JSON with an embedded attestation
+// of some type or other.
+func UntypedProvenanceFromSignature(sig oci.Signature) (Attestation, error) {
+	payload, err := PayloadFromSig(sig)
+	if err != nil {
+		return nil, err
+	}
+
+	embedded, err := decodedPayload(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	var statement UntypedAttestation
+	if err := json.Unmarshal(embedded, &statement); err != nil {
+		return nil, AT002.CausedBy(err)
+	}
+
+	signatures, err := createEntitySignatures(sig, payload)
+	if err != nil {
+		return nil, AT005.CausedBy(err)
+	}
+
+	// Fix string predicate if needed
+	if statement.predicateIsString() {
+		fixedPredicate, err := statement.Predicate()
+		if err != nil {
+			return nil, AT002.CausedBy(err)
+		}
+		statement[predicateKey] = fixedPredicate
```

Comment:
Would it make sense to store this in a field of `untypedProvenance`?

Reply:
Not sure I follow, do you mean to suggest we have the existing `statement` field (which might be "typed") as well as another `untypedStatement` field which is always un"typed"?

Reply:
Relates to https://github.com/enterprise-contract/ec-cli/pull/975#discussion_r1312807789, for having the predicate type as a field on `untypedProvenance` and not modify the `statement`, i.e. provide it to the rules as is...